spring-kafka/src/main/java/org/springframework/kafka/annotation/ResolvableType.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2014-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.annotation;
+
+import org.springframework.util.MimeTypeUtils;
+
+import java.lang.annotation.*;
+
+/**
+ * Used to define the content type of {@code Message} with {@code MediaMessageResolver}.
+ *
+ * @author Scruel Tao
+ */
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.PARAMETER})
+public @interface ResolvableType {
```

Comment:
There is already a `org.springframework.core.ResolvableType`.
So, this one is going to confuse.
Plus it doesn't look like we really resolve a type, but talk more about a `content-type`.

I'm not fully sure that I follow with the goal of this feature, but probably better to have it as an `@ExpectedContentType` since there are many `ContentType` classes around 😸 .
I would say this is similar to `consumes` on the `@RequestMapping`, but you really have here a fallback to the expected if no one provided in the record.
Anyway I'd like to understand better what is going on.
Looks like you want to have a common composite converter and choose an appropriate one according the `content-type` header or fallback.
How about have an attribute on the `@KafkaListener` instead? Kinda `fallbackContentType`?
Populate it to the message (if needed) and still rely on that composite ability to choose an expected converter.

Just thinking aloud...

Thanks

Reply:
Thanks, Artem, for your perspective; I agree coercing the content type in the listener adapter via a `@KafkaListener` property would be better than polluting the method signature.

Reply:
> Thanks, Artem, for your perspective; I agree coercing the content type in the listener adapter via a `@KafkaListener` property would be better than polluting the method signature.

In spring, annotation on parameters is very common, I think it can not treat as polluting the method signature, for example `@RequestBody`.

Reply:
Yes, but we already have enough param annotations to pollute enough. 
And your idea makes me think that this kind of `fallbackContentType` should go to the existing `@Payload` annotation which we rely in Spring Kafka as well.
But again: your current goal could be achieved with an explicit converter for the particular `@KakfaListener`. The rest I believe is already covered by the composition if `contentType` header is present.

Reply:
Hah, `@Payload` annotation is in `org.springframework.messaging` package but not this project, I treat the `@MediaMessage` as an enhanced annotation for `spring-kafka`.
I know it was covered if `contentType` header is present, but sometimes, you can not control the producers for which message they will send.

Reply:
> Thanks; it needs some docs; probably in this section https://github.com/spring-projects/spring-kafka/blob/main/spring-kafka-docs/src/main/asciidoc/kafka.adoc#kafkalistener-annotation
> 
> Also, maybe a "how to use" here https://docs.spring.io/spring-kafka/docs/current/reference/html/#tips-n-tricks

I will fulfill the docs after you make your decision.

Reply:
> Thanks, Artem, for your perspective; I agree coercing the content type in the listener adapter via a `@KafkaListener` property would be better than polluting the method signature.

Would you need me also to implement this idea?

Reply:
Thanks, I agree that the other projects have the rights to benefit from this.
> they still need to support an infrastructure for converter composition and so on.
I do want to add this in default configuration, but maybe we should consider how to implement this first, then consider should this support feature as a default feature.
I will wait for your decision, feel free to comment your ideas.